src/SoapCore/ServiceBodyWriter.cs

```diff
@@ -20,151 +22,173 @@ public class ServiceBodyWriter : BodyWriter
 		private readonly object _result;
 		private readonly Dictionary<string, object> _outResults;
 
-		public ServiceBodyWriter(SoapSerializer serializer, OperationDescription operation, string resultName, object result, Dictionary<string, object> outResults) : base(isBuffered: true)
+		public ServiceBodyWriter(SoapSerializer serializer, OperationDescription operation, object result, Dictionary<string, object> outResults) : base(isBuffered: true)
 		{
 			_serializer = serializer;
 			_operation = operation;
 			_serviceNamespace = operation.Contract.Namespace;
 			_envelopeName = operation.Name + "Response";
-			_resultName = resultName;
+			_resultName = operation.ReturnName;
 			_result = result;
-			_outResults = outResults;
+			_outResults = outResults ?? new Dictionary<string, object>();
 		}
 
 		protected override void OnWriteBodyContents(XmlDictionaryWriter writer)
 		{
-			// do not wrap old-style single element response into additional xml element for xml serializer
-			var needResponseEnvelope = _serializer != SoapSerializer.XmlSerializer || _result == null || (_outResults != null && _outResults.Count > 0) || !_operation.IsMessageContractResponse;
+			switch (_serializer)
```

Comment:
Instead of checking the cases in-line I have opted to split the xml/data contract logic internally at a higher level.

This has resulted in simpler logic to follow if you are concerned about a single path and allows me to isolate the serialization change (to not unexpected impact DataContractSerializer).

Unfortunately it does duplicate some logic.